bundles/org.openhab.core/src/main/java/org/openhab/core/items/GenericItem.java

```diff
@@ -251,18 +251,23 @@ protected void notifyListeners(final State oldState, final State newState) {
         Set<StateChangeListener> clonedListeners = null;
         clonedListeners = new CopyOnWriteArraySet<>(listeners);
         ExecutorService pool = ThreadPoolManager.getPool(ITEM_THREADPOOLNAME);
-        for (final StateChangeListener listener : clonedListeners) {
-            pool.execute(() -> {
-                try {
-                    listener.stateUpdated(GenericItem.this, newState);
-                    if (newState != null && !newState.equals(oldState)) {
-                        listener.stateChanged(GenericItem.this, oldState, newState);
+        try {
+            final boolean stateChanged = newState != null && !newState.equals(oldState);
+            clonedListeners.forEach(listener ->
+                pool.execute(() -> {
+                    try {
+                        listener.stateUpdated(GenericItem.this, newState);
+                        if (stateChanged) {
+                            listener.stateChanged(GenericItem.this, oldState, newState);
+                        }
+                    } catch (Exception e) {
+                        logger.warn("failed notifying listener '{}' about state update of item {}: {}", listener,
+                                GenericItem.this.getName(), e.getMessage(), e);
                     }
-                } catch (Exception e) {
-                    logger.warn("failed notifying listener '{}' about state update of item {}: {}", listener,
-                            GenericItem.this.getName(), e.getMessage(), e);
-                }
-            });
+                }));
+        } catch (Exception e) {
```

Comment:
Changed to IAE. WDYT about making `compareTo` a `synchronized` method?

Reply:
In general I am not a big fan of `synchronized` for this method because we cannot be sure the result is still the same after the synchronization was left. If we want to force that we have to snyc the whole block - the compared objects too. We should make sure the values used for comparison are immutable and preferable final.

Reply:
The problem is not our code. The problem is that https://github.com/unitsofmeasurement/uom-se/blob/b9b4d770e25a245492d98909529e51961edb7bc3/src/main/java/tec/uom/se/unit/ProductUnit.java#L75 is not as final as it looks (the array itself is final, it’s elements are not) and during comparison the elements are changed https://github.com/unitsofmeasurement/uom-se/blob/b9b4d770e25a245492d98909529e51961edb7bc3/src/main/java/tec/uom/se/unit/ProductUnit.java#L279 which probably causes a problem if the same `Unit` is compared a second time while the first comparison is not finished. That explains why moving the comparison solves our problem and why it only happens under high load conditions when a lot of scheduled comparison are executed.

I don‘t see an easy way to solve this (except changing the array to a thread-safe collection-type).

Reply:
Your current change is a good step and I will merge it for now. I think we can synchronize the comparison later if needed.

Maybe @keilw is interested in this topic.